golem/resources/debit_note/debit_note.py

```diff
@@ -45,58 +45,92 @@ def activity(self) -> "Activity":
     async def get_status(self) -> PayDocumentStatus:
         return PayDocumentStatus(str((await self.get_data()).status).upper())
 
-    async def get_previous_debit_note_data(self, timestamp: datetime) -> Optional[models.DebitNote]:
-        """Get latest debit note data from before given `timestamp`."""
-        debit_notes_data = [(await dn.get_data()) for dn in self.parent.debit_notes]
-        debit_notes_data.sort(key=lambda dn: dn.timestamp, reverse=True)
-        for dn in debit_notes_data:
-            if dn.timestamp < timestamp:
-                return dn
+    async def get_previous_debit_note(self) -> Optional["DebitNote"]:
+        """Get previous debit note."""
+        for debit_note in sorted(
+            self.activity.debit_notes, key=lambda dn: dn.created_at, reverse=True
+        ):
+            if debit_note.created_at < self.created_at:
+                return debit_note
         return None
```

Comment:
this should be more efficient if you just need one value as it omits the sorting:

```suggestion
        try:
            return max([dn for self.activity.debut_notes if dn.created_at < self.created_at, key: lambda dn: dn.created_at)
        except ValueError:
            return None
```

obviously, it could even be done in one pass but it wouldn't look as neat

Reply:
Changed, I needed to add `# type: ignore[union-attr]` as mypy was complaining
`error: Item "None" of "DebitNote | None" has no attribute "created_at"  [union-attr]`